### ISSUE TITLE
questions about preprocessing

### ISSUE THREAD

**Issue by stas-polukeev** [KelianB/SPARK#12]
Hey, thanks for the great work! really loved your approach described in the paper!
However, I wanted to ask about the steps you took to preprocess your data. In the paper's appendix I haven't found detailed description in the preprocessing pipeline. I just launched the MonoFaceCapture preprocessing, but if it's possible i would like what steps you took exactly in the spark paper, as MonoFaceCapture seems to be to broad.
Thanks in advance!

**Comment by stas-polukeev** [comment]
Yes, thank  you, that was well detailed!
Also, I haven't fully understood, whether it's possible to do without cropping, as it needs to have some kind of face detector during inference. I did it with yolov5-face for some speed, however, training without it would be more convenient, as the pipeline is more straight forward. I saw the render mode "full" instead of "crop" but haven't fully understand whether its what I'm thinking of. Is this the way to get without cropping or the results will be significantly worse?

**Comment by KelianB** [comment]
To clarify:
- You need to have 512x512 crops for the avatar training stage of SPARK. However, these crops don't need to be very tight around the face, and I found it best to use a relatively wide **fixed** window (which is rescaled to 512x512 after cropping). These are the "full" crops.
- For inference through the pretrained feedforward trackers (DECA, EMOCA, SMIRK), you need a much tighter crop as this is what they were trained with. So for the tracker adaptation stage of SPARK, we compute a new crop from the landmark positions. **Without this, performance highly degrades as we are outside of the pretraining distribution**.
- When visualizing the results, we can either use render mode "full" (crops are fixed, wider, and include the full head and some margin around) or render mode "crop" (tighter landmarks-based crops). Either way the tighter crops are used for inference, the option is just for visualization.

Unfortunately I don't see an easy way to do without a face detector during inference.

**Comment by stas-polukeev** [comment]
@KelianB Thanks for the clarification!
The second, "tight" crops, as you say, are they saved somewhere or are they computed in place somewhere in the code? If they are just the face, which is somewhat rectangular do you pad the rectangular crop or warp it to a square? I mimicked the preprocessing "wide" crops and the results seem to be poor during inference.

**Comment by KelianB** [comment]
The tight crops are computed on-the-fly in the CropDataset class, they are not saved to disk.
It simply takes the min/max FAN landmarks coordinates to compute a rectangle box, makes that box a square (same center coordinate with size=(width+height)/2), scales up the box size by 1.35 and then moves it to ensure it's fully contained within the source image.

If you have access to the full sequence (ie not a real-time application) I also advise applying some gaussian filtering on the box coordinates, otherwise the crops will have a bit of high-frequency jitter that is visible in the final outputs.

This is all implemented in `TrackerAdaptation/adapt/crop_dataset.py`.

**Comment by stas-polukeev** [comment]
@KelianB Thanks for a detailed reply and sorry if I ask too many questions. I am trying to make the flame parameter generation applicable in real-time scenario, but I am a little bit confused with the image preprocessing. Assuming, I've got the "tight" crops looking almost  identical, i suppose it could not be the problem. I keep getting the results, which when rendered produce little to no movement and seem to be looking just like white noise. I suspect that it can be due to the fact, that i preprocess images differently before passing them into the SMIRKEncoder to get flame params. 
As far I could deduce that you do the following:
1. load the image from image path with load_image() function
2. convert it with srgb_to_rgb
3. crop it "tightly" 
4. convert the crop with rgb_to_srgb
5. permute channels for torch compatibility
6. feed it to the encoder

why do we need srgb convertions and have i missed some important steps in image processing? Thanks for your patience

**Comment by stas-polukeev** [comment]
and is eye tracking possible with spark? as far as i can see, the models mostly have their eyes fixed

**Comment by KelianB** [comment]
The srgb conversion is for the avatar reconstruction, it's not needed for inference of the trained tracker. The srgb to rgb to srgb conversion is just a quirk of the implementation. If the rendered geometry looks like white noise there might be a scaling issue in the pose? Like a wrong z coordinate?
It's hard to help without more information, but if it's just a preprocessing issue at worse you should get bad poses/expressions, but you should still be seeing the FLAME head.

**Comment by stas-polukeev** [comment]
Excuse me, I was a bit misleading.
What i meant is when i infer the sequence of acquired expressions and poses I see a head in a canonical position, but all it does is shake a little as if the distribution of expr/pose params is white noise with little variance.  The shaking is really little, and the face is almost not deforming.
However, during test_fn() in your code the results are rendered fine, so I guessed that the problem could be in the image processing before passing it to the encoder (SMIRK in my case), as the outputs of the model all have low variance, as if the decoder is receiving something it has not seen
So to be clear, the basic canonical geometry behaves as normal, but it is all inherited from the trained head model
The encoded poses and expressions are not interpreted correctly by the flame itself, however, I've checked that Flame parameters and MultiFLARE parameters (the decoder in my case) are the same as I have turned off both remeshing and deformer net

The crops also look the same as those used by the CropDataset, so I don't think that the cropping might be the problem.

I did copy the processing of parameters from the decoder that it does before passing it into flame. As far as I could see if remeshing is off and the deformer returns the original flame the inference should be identical using FLAME. The version of FLAME is matched.
going from basic cv2.imread(), cv2.BGR2RGB did not work, as did not the code I took from the dataset classes with load_img() and mentioned above srgb_to_rgb() funcitions

**Comment by KelianB** [comment]
Can you try using the pre-trained SMIRK weights and see what happens?

After cropping, do you resize to 224x224?

**Comment by stas-polukeev** [comment]
@KelianB sorry for time wasted, but the problem was actually that the frames of the video were stored in a folder, and when the model read them with os.listdir() it did not sort it, so it just return poses of random frames. After I sorted, the results are fine. However, eyes seem to be more open than they should be (if i use default FLAME and not multiflare as a decoder). Do you think, the reason could be the way SMIRK (which i use as a backbone) handles eyelids (as far as I understand, they are processed separately)?

**Comment by KelianB** [comment]
Glad you could fix your issue!
Your problem with eyes is definitely because of the separate handling of eyelids by SMIRK. You can replicate the same thing in your FLAME, it's pretty much a blendshape, you just need to get the eyelid masks from SMIRK. You can check how the MultiFLARE decoder does it (face_decoder.py L186).

**Comment by stas-polukeev** [comment]
@KelianB thanks for the suggestion, I'll probably try it!
